Lib/test/datetimetester.py

```diff
@@ -2334,6 +2390,129 @@ def newmeth(self, start):
         self.assertEqual(dt2.newmeth(-7), dt1.year + dt1.month +
                                           dt1.second - 7)
 
+    def test_fromisoformat_datetime(self):
+        # Test that isoformat() is reversible
+        base_dates = [
+            (1, 1, 1),
+            (1000, 2, 14),
+            (1900, 1, 1),
+            (2004, 11, 12),
+            (2004, 4, 3),
+            (2017, 5, 30)
+        ]
+
+        base_times = [
+            (0, 0, 0, 0),
+            (0, 0, 0, 241000),
+            (0, 0, 0, 234567),
+            (23, 59, 47),
+            (12, 30, 45, 234567)
+        ]
+
+        separators = [
+            ' ', 'T', '\u007f',     # 1-bit widths
+            '\u0080', 'ʁ',          # 2-bit widths
+            'ᛇ', '時',               # 3-bit widths
+            '🐍'                     # 4-bit widths
+        ]
+
+        tzinfos = [None, timezone.utc,
+                   timezone(timedelta(hours=-5)),
+                   timezone(timedelta(hours=2)),
+                   timezone(timedelta(hours=6, minutes=27))]
+
+        dts = [self.theclass(*date_tuple, *time_tuple, tzinfo=tzi)
+               for date_tuple in base_dates
+               for time_tuple in base_times
+               for tzi in tzinfos]
+
+        for dt in dts:
+            for sep in separators:
```

Comment:
6 × 5 × 8 × 5 is 1200 tests. Perhaps just one test for each case (6 + 5 + 8 + 5) instead?

Reply:
This is a good point. I translated these tests from hypothesis where the space is sampled sparsely. I think I can refactor out at least the separator dimension and probably reduce the dimensionality of some of the remaining tests. I'll look at it tomorrow.

Reply:
I refactored out the separator tests and reduced some of the space that is sampled. I think most of the variation in the date and time portions separately is covered by the respective date and time parser functions. I think we're down to ~160 tests.